### ISSUE TITLE
Tweak request to make integration with webpack-subresource-integrity easy

### ISSUE THREAD

**Issue by SPSpwetter** [ampedandwired/html-webpack-plugin#330]
I got sub resource integrity checking working using this slightly hacky code:

https://github.com/waysact/webpack-subresource-integrity/issues/3

The hacky part is that I hard coded my file name template (chunk.[name].bundle.js] in order to build the key for compilation.assets. If `htmlWebpackPlugin.files.chunks[chunk].assetKey` had the asset key, that would make it less hacky. Or maybe i'm doing it wrong.

**Comment by jantimon** [comment]
Well as `inject:true` is the default you might want to go with it and add a hook to alter the scripts.

**Comment by jantimon** [comment]
Take a look at the `'html-webpack-plugin-before-html-generation'` it will pass you all information you need.

**Comment by SPSpwetter** [comment]
Ok, I'm willing to do that work, but you have to understand that I didn't understand what you said having only started working with webpack a month ago in an effort to migrate a Makefile+cat build to something more reasonable. Could you spell out step by step a bit more what you think I should do?

**Comment by SPSpwetter** [comment]
I started down the path of what I thought you were asking me to do, but it was a 1-line change to the plugin to squirrel away the base name of the file when building the chunk hash, and I couldn't figure out how to debug a plugin very effectively never having written a webpack plugin before, though the README.md on this project was a good start. 

So I submitted the above pull request instead. With this pull request, adding integrity checks or anything that can be read from compilation.assets is just:

`integrity = "<%= compilation.assets[htmlWebpackPlugin.files.chunks[chunk].basename].integrity %>"`

In short: 

You should accept the above pull request because:

1. It's benign, and simple. 
2. It makes a whole collection of things that might want to extend the html output easier, if they can just annotate compilation.assets.

**Comment by jantimon** [comment]
Why don't you use a hook?

Like https://github.com/jantimon/favicons-webpack-plugin/blob/master/index.js#L57-L62

**Comment by SPSpwetter** [comment]
Then someone needs 3 plugins+template to implement SRI:

1. The SRI plugin. 
2. HTML webpack
3. My hook plugin. 
4. Custom templates. 

I figured having the base name available was a generally useful enhancement that would enable people to do whatever they needed in a template by writing a plugin to enhance the compilation object, where they could pick it up later. 

Another alternative would be if the path module was available inside templates, path.basename() would work with slightly more complex templates.

If you want it to work with injection, I could also modify the injection to see if the integrity value is present for an asset and modify the injection tag there. Then no template is necessary.

**Comment by jantimon** [comment]
Some reasons why this is not my preferred way to solve this:
1) `assets.chunks` are legacy from 1.0 and might become deprecated as we have `assets.js` and `assets.css` now - so I would rather not extend them any further.
2) `inject:true` is the default behaviour and it would not work with your approach

With your solution it would take a `inject:false`, a custom template, the sri plugin and the html webpack plugin.

I would propose to add a hook to the SRI plugin then we could have a configuration like this to get it working:

```js
plugins: [
        new WebpackHtmlPlugin(),
        new SriPlugin(['sha256', 'sha384']),
    ]
```

What do you think?

**Comment by SPSpwetter** [comment]
Simpler would definitely be better, it would be best to just have two plugins, but I wasn't sure how to mutate the <script> tags after they'd already been injected.  In the sample you sent me, you're doing a search/replace on the <head tag, but that seems kind of hacky and error prone to me.

It would have to find the already added script tags, parse them to get the path, look them up in the compilation object, then write out the integrity attribute.  Ick. 

If you can give me more of a hint on a way to change the script tags more elegantly, I can look into adding a hook to the SRI plugin instead.

Thinking about it, a cleaner way would be to allow hooks to add additional attributes prior to injection to assets.js and assets.css in assets.attributes[]. The injection code could then loop over any any additional attributes and inject them. It would be slightly more complexity on line 454 and 459 of index.js but not too bad. 

```
attr = assets.additions[scriptPath]
if (attr)
{
additionalAttributes = Object.keys(attr).map( function(key) {key + '="'+attr[key]+"'"'}).join(" ")
}
```

Do you like that?

**Comment by jantimon** [comment]
Oh I like that idea - checkout #345

**Comment by SPSpwetter** [comment]
Oh, great! I'll work on the SRI plugin hook on the assumption this will go in. Much cleaner than my idea in the details.

**Comment by jantimon** [comment]
Cool please let me know if it doesn't work properly

**Comment by SPSpwetter** [comment]
SRI Plugin mods built, see PR: https://github.com/waysact/webpack-subresource-integrity/pull/5

**Comment by jantimon** [comment]
Integrated in webpack-subresource-integrity 0.4.0
